mission/FSM/docking_cpp/src/docking.cpp

```diff
@@ -0,0 +1,574 @@
+                    feedback->feedback.z);
+    }
+}
+
+ConvergeDockingStationState::ConvergeDockingStationState()
+    : yasmin_ros::ActionState<ReferenceFilterWaypoint>(
+          "/reference_filter",
+          std::bind(&ConvergeDockingStationState::create_goal_handler,
+                    this,
+                    _1),
+          std::bind(&ConvergeDockingStationState::response_handler,
+                    this,
+                    _1,
+                    _2),
+          std::bind(&ConvergeDockingStationState::print_feedback,
+                    this,
+                    _1,
+                    _2)) {};
+
+ReferenceFilterWaypoint::Goal ConvergeDockingStationState::create_goal_handler(
+    std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
+    auto goal = ReferenceFilterWaypoint::Goal();
+
+    PoseStamped dock_pose = blackboard->get<PoseStamped>("dock_pose");
+    goal.goal = dock_pose;
+
+    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Goal sent to action server:");
+    RCLCPP_INFO(rclcpp::get_logger("rclcpp"),
+                "  Position: x = %f, y = %f, z = %f\n",
+                goal.goal.pose.position.x, goal.goal.pose.position.y,
+                goal.goal.pose.position.z);
+
+    return goal;
+}
+
+std::string ConvergeDockingStationState::response_handler(
+    std::shared_ptr<yasmin::blackboard::Blackboard> blackboard,
+    ReferenceFilterWaypoint::Result::SharedPtr response) {
+    RCLCPP_INFO(rclcpp::get_logger("rclcpp"),
+                "Response received from action server:");
+    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "  Success: %s\n",
+                response->success ? "true" : "false");
+
+    blackboard->set<bool>("has_finished_converging", response->success);
+    if (blackboard->get<bool>("has_finished_converging")) {
+        return yasmin_ros::basic_outcomes::SUCCEED;
+    } else {
+        return yasmin_ros::basic_outcomes::CANCEL;
+    }
+}
+
+void ConvergeDockingStationState::print_feedback(
+    std::shared_ptr<yasmin::blackboard::Blackboard> blackboard,
+    std::shared_ptr<const ReferenceFilterWaypoint::Feedback> feedback) {
+    Pose current_pose = Pose();
+    current_pose.position.x = feedback->feedback.x;
+    current_pose.position.y = feedback->feedback.y;
+    current_pose.position.z = feedback->feedback.z;
+
+    blackboard->set<Pose>("current_pose", current_pose);
+
+    if (blackboard->get<bool>("print_debug")) {
+        RCLCPP_INFO(rclcpp::get_logger("rclcpp"),
+                    "Current position: x = %f, y = %f, z = %f\n",
+                    feedback->feedback.x, feedback->feedback.y,
+                    feedback->feedback.z);
+    }
+}
+
+std::string DockedState(
+    std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
+    std::chrono::duration<double> elapsed_time = std::chrono::seconds(0);
+    std::chrono::time_point<std::chrono::system_clock> start_time =
+        std::chrono::system_clock::now();
```

Comment:
IIRC system_clock is not guaranteed to be monotonic. Since you rely on it to get out of a blocking while-loop, you might want to consider `std::chrono::steady_clock`